File: library/src/main/java/me/shkschneider/skeleton/helper/LocaleHelper.java
package me.shkschneider.skeleton.helper;

import android.support.annotation.NonNull;
import android.support.v4.os.LocaleListCompat;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class LocaleHelper {

    protected LocaleHelper() {
        // Empty
    }

    public static List<Locale> locales() {
        final List<Locale> locales = new ArrayList<>();
        final LocaleListCompat localeListCompat = LocaleListCompat.getDefault();
        for (int i = 0; i < localeListCompat.size(); i++) {
            locales.add(localeListCompat.get(i));
        }
        return locales;
    }

    public static Locale locale() {
        return LocaleListCompat.getDefault().get(0);
    }

    @Deprecated
    public static String language() {
        return language(locale());
    }

    public static String language(@NonNull final Locale locale) {
        return locale.getDisplayLanguage();
    }

    @Deprecated
    public static String language2() {
        return language2(locale());
    }

    public static String language2(@NonNull final Locale locale) {
        return locale.getLanguage();
    }

    @Deprecated
    public static String language3() {
        return language3(locale());
    }

    public static String language3(@NonNull final Locale locale) {
        return locale.getISO3Language();
    }

    @Deprecated
    public static String country() {
        return country(locale());
    }

    public static String country(@NonNull final Locale locale) {
        return locale.getDisplayCountry();
    }

    @Deprecated
    public static String country2() {
        return country2(locale());
    }

    public static String country2(@NonNull final Locale locale) {
        return locale.getCountry();
    }

    @Deprecated
    public static String country3() {
        return country3(locale());
    }

    public static String country3(@NonNull final Locale locale) {
        return locale.getISO3Country();
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/helper/LocaleHelper.kt
package me.shkschneider.skeleton.helper

import android.content.Context
import android.content.ContextWrapper
import android.content.res.Configuration
import android.os.Build
import android.support.v4.os.LocaleListCompat
import java.util.*

object LocaleHelper {

    object Application {

        // attachBaseContext() -> super.attachBaseContext(switch())
        fun switch(context: Context?, locale: Locale): ContextWrapper {
            context ?: return ContextWrapper(null)
            with(context.resources) {
                val configuration = Configuration()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                    configuration.setLocale(locale)
                } else {
                    @Suppress("DEPRECATION")
                    configuration.locale = locale
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                    return ContextWrapper(context.createConfigurationContext(configuration))
                } else {
                    @Suppress("DEPRECATION")
                    updateConfiguration(configuration, displayMetrics)
                    return ContextWrapper(context)
                }
            }
        }

        fun locale(context: Context): Locale {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                return context.resources.configuration.locales[0]
            } else {
                @Suppress("DEPRECATION")
                return context.resources.configuration.locale
            }
        }

    }

    object Device {

        fun locales(): List<Locale> {
            val localeListCompat = LocaleListCompat.getDefault()
            return (0 until localeListCompat.size()).map { localeListCompat.get(it) }
        }

        fun locale(): Locale {
            return locales()[0]
        }

    }

    fun language(locale: Locale): String {
        return locale.displayLanguage
    }

    fun language2(locale: Locale): String {
        return locale.language
    }

    fun language3(locale: Locale): String {
        return locale.isO3Language
    }

    fun country(locale: Locale): String {
        return locale.displayCountry
    }

    fun country2(locale: Locale): String {
        return locale.country
    }

    fun country3(locale: Locale): String {
        return locale.isO3Country
    }

}
